src/types/plugin.ts

```diff
@@ -2,9 +2,9 @@ import { CfnWafRuleStatement, IntrinsicFunction } from './cloudFormation';
 
 export type AppSyncConfig = {
```

Comment:
https://github.com/sid88in/serverless-appsync-plugin/pull/597/files#r1182585171
> If most of the parameters are ignored when using apiId you should define type like this to avoid misusage of the config, it can be nice for those who use this type in their serverless.ts:
> 
> ```
> export type BaseAppSyncConfig = {
>   dataSources: Record<string, DataSourceConfig>;
>   resolvers: Record<string, ResolverConfig>;
>   pipelineFunctions: Record<string, PipelineFunctionConfig>;
>   substitutions?: Substitutions;
>   caching?: CachingConfig;
> };
> 
> export  type NewAppSyncConfig = BaseAppSyncConfig & {
>   name: string;
>   schema: string[];
>   authentication: Auth;
>   additionalAuthentications: Auth[];
>   domain?: DomainConfig;
>   apiKeys?: Record<string, ApiKeyConfig>;
>   xrayEnabled?: boolean;
>   logging?: LoggingConfig;
>   waf?: WafConfig;
>   tags?: Record<string, string>;
> };
> 
> export type ExistingAppSyncConfig = BaseAppSyncConfig & {
>   apiId: string | IntrinsicFunction;
> };
> 
> export type AppSyncConfig = NewAppSyncConfig | ExistingAppSyncConfig;
> ```

https://github.com/sid88in/serverless-appsync-plugin/pull/597/files#r1223318608
> I agree on this and it joins what I commented earlier. The same should happen in the validation json schema. I would use something like a union.
> 
> ```ts
> export type BaseAppSyncConfig = {
>   dataSources: Record<string, DataSourceConfig>;
>   resolvers: Record<string, ResolverConfig>;
>   pipelineFunctions: Record<string, PipelineFunctionConfig>;
>   substitutions?: Substitutions;
>   caching?: CachingConfig;
> };
> 
> export  type FullAppSyncConfig = BaseAppSyncConfig & {
>   name: string;
>   schema: string[];
>   authentication: Auth;
>   additionalAuthentications: Auth[];
>   domain?: DomainConfig;
>   apiKeys?: Record<string, ApiKeyConfig>;
>   xrayEnabled?: boolean;
>   logging?: LoggingConfig;
>   waf?: WafConfig;
>   tags?: Record<string, string>;
> };
> 
> export  type SharedAppSyncConfig = BaseAppSyncConfig & {
>   apiId: string;
> };
> 
> export type AppSyncConfig = FullAppSyncConfig | SharedAppSyncConfig
> ```
> 
> (not tested, might need adjustments)